### PR TITLE
Implemented qlaunch version of the controller applet

### PR DESCRIPTION
**Pull request by Macj0rdan**
With a lot of help from @german77 I managed to implement the qlaunch version of the controller applet. 
It's not super useful for now but it's a good start to get this running in games.

Small clip of it working

### CODE REVIEW

File: src/core/hle/service/am/am.cpp [hunk 1]
```diff
@@ -97,6 +101,17 @@ void IWindowController::GetAppletResourceUserId(HLERequestContext& ctx) {
     rb.Push<u64>(process_id);
 }
 
+void IWindowController::GetAppletResourceUserIdOfCallerApplet(HLERequestContext& ctx) {
+
```
Comment (german77) [inline src/core/hle/service/am/am.cpp:105]: Remove this empty line

File: src/core/hle/service/am/am.cpp [hunk 2]
```diff
@@ -1666,6 +1684,25 @@ void ILibraryAppletSelfAccessor::GetLibraryAppletInfo(HLERequestContext& ctx) {
     rb.PushRaw(applet_info);
 }
 
+void ILibraryAppletSelfAccessor::GetMainAppletIdentityInfo(HLERequestContext& ctx) {
+    struct AppletIdentityInfo {
+        Applets::AppletId applet_id;
+        INSERT_PADDING_BYTES(0x4);
+        u64 application_id;
+    };
```
Comment (german77) [inline src/core/hle/service/am/am.cpp:1692]: Add assert on struct size

File: src/core/hle/service/am/am.cpp [hunk 3]
```diff
@@ -1737,6 +1774,61 @@ void ILibraryAppletSelfAccessor::PushInShowAlbum() {
     queue_data.emplace_back(std::move(settings_data));
 }
 
+void ILibraryAppletSelfAccessor::PushInShowController() {
+
```
Comment (german77) [inline src/core/hle/service/am/am.cpp:1778]: ditto. Don't leave empty lines after the function header

File: src/core/hle/service/am/am.cpp [hunk 4]
```diff
@@ -1737,6 +1774,61 @@ void ILibraryAppletSelfAccessor::PushInShowAlbum() {
...
+    const auto hid = system.ServiceManager().GetService<Service::HID::Hid>("hid");
+    const auto applet_resource = hid->GetAppletResource();
+    Applets::CommonArguments common_args = {
+        .arguments_version = Applets::CommonArgumentVersion::Version3,
+        .size = Applets::CommonArgumentSize::Version3,
+        .library_version = 1,
```
Comment (german77) [inline src/core/hle/service/am/am.cpp:1784]: ControllerAppletVersion::Version8

File: src/core/hle/service/am/am.cpp [hunk 5]
```diff
@@ -1737,6 +1774,61 @@ void ILibraryAppletSelfAccessor::PushInShowAlbum() {
...
+
+    Applets::ControllerSupportArgNew user_args = {
+        .header = {.player_count_min = 1,
+                   .player_count_max = 4,
+                   .enable_take_over_connection = true,
+                   .enable_left_justify = false,
+                   .enable_permit_joy_dual = true,
+                   .enable_single_mode = false,
+                   .enable_identification_color = false},
+        .identification_colors =
```
Comment (german77) [inline src/core/hle/service/am/am.cpp:1798]: identification colors are disabled. 
```suggestion
        .identification_colors = {},
```

File: src/core/hle/service/am/am.cpp [hunk 6]
```diff
@@ -1737,6 +1774,61 @@ void ILibraryAppletSelfAccessor::PushInShowAlbum() {
...
+                Applets::IdentificationColor{1, 0, 0, 0},
+            },
+        .enable_explain_text = false,
+        .explain_text = {},
+    };
+
+    Applets::ControllerSupportArgPrivate private_args = {
+        .arg_private_size = sizeof(Applets::ControllerSupportArgPrivate),
+        .arg_size = sizeof(Applets::ControllerSupportArgNew),
+        .flag_0 = true,
```
Comment (german77) [inline src/core/hle/service/am/am.cpp:1809]: rename it to. is_home_menu

File: src/core/hle/service/am/am.cpp [hunk 7]
```diff
@@ -1737,6 +1774,61 @@ void ILibraryAppletSelfAccessor::PushInShowAlbum() {
...
+    std::vector<u8> user_args_data(sizeof(user_args));
+
+    std::memcpy(common_args_data.data(), &common_args, sizeof(common_args));
+    std::memcpy(private_args_data.data(), &private_args, sizeof(private_args));
+    std::memcpy(user_args_data.data(), &user_args, sizeof(user_args));
+
+    queue_data.emplace_back(std::move(common_args_data));
+    queue_data.emplace_back(std::move(private_args_data));
+    queue_data.emplace_back(std::move(user_args_data));
+} // namespace Service::AM
```
Comment (german77) [inline src/core/hle/service/am/am.cpp:1830]: Namespace doesn't end here?

File: src/core/hle/service/am/applets/applet_controller.cpp [hunk 8]
```diff
@@ -5,6 +5,7 @@
 #include <cstring>
 
 #include "common/assert.h"
+#include "common/hex_util.h"
```
Comment (german77) [inline src/core/hle/service/am/applets/applet_controller.cpp:8]: Unnecessary include

File: src/core/hle/service/am/applets/applet_controller.h [hunk 9]
```diff
@@ -34,7 +34,7 @@ enum class ControllerAppletVersion : u32_le {
 
 enum class ControllerSupportMode : u8 {
     ShowControllerSupport,
-    ShowControllerStrapGuide,
+    ShowControllerStrapGuide, // Show the Strap guide for joycons ?
```
Comment (german77) [inline src/core/hle/service/am/applets/applet_controller.h:37]: Yes this is the strap guide. No need to say the same on the comment.

File: src/core/hle/service/hid/controllers/npad.cpp [hunk 10]
```diff
@@ -344,6 +344,7 @@ void Controller_NPad::InitNewlyAddedController(Core::HID::NpadIdType npad_id) {
         controller.device->SetPollingMode(Core::HID::EmulatedDeviceIndex::AllDevices,
                                           Common::Input::PollingMode::Active);
     }
+
```
Comment (german77) [inline src/core/hle/service/hid/controllers/npad.cpp:347]: Unrelated change

File: src/core/hle/service/hid/controllers/npad.cpp [hunk 11]
```diff
@@ -1724,4 +1725,18 @@ const Controller_NPad::SixaxisParameters& Controller_NPad::GetSixaxisState(
...
+Controller_NPad::AppletDetailedUiType Controller_NPad::GetAppletDetailedUiType(
+    Core::HID::NpadIdType npad_id) {
+
+    auto controller = GetControllerFromNpadIdType(npad_id);
+    auto shared_memory = controller.shared_memory;
+    Service::HID::Controller_NPad::AppletFooterUiType applet_footer_type =
+        shared_memory->applet_footer_type;
+
+    Controller_NPad::AppletDetailedUiType detailed_ui_type{
+        .footer = applet_footer_type,
```
Comment (german77) [inline src/core/hle/service/hid/controllers/npad.cpp:1737]: ui_variant needs to be initialized

File: src/core/hle/service/hid/controllers/npad.h [hunk 12]
```diff
@@ -529,7 +539,6 @@ class Controller_NPad final : public ControllerBase {
         const Core::HID::VibrationDeviceHandle& device_handle) const;
     NpadControllerData& GetControllerFromNpadIdType(Core::HID::NpadIdType npad_id);
     const NpadControllerData& GetControllerFromNpadIdType(Core::HID::NpadIdType npad_id) const;
-
```
Comment (german77) [inline src/core/hle/service/hid/controllers/npad.h:532]: Unrelated change

File: src/core/hle/service/hid/hid.cpp [hunk 13]
```diff
@@ -1004,8 +1005,9 @@ void Hid::SetSupportedNpadStyleSet(HLERequestContext& ctx) {
 
     const auto parameters{rp.PopRaw<Parameters>()};
 
-    applet_resource->GetController<Controller_NPad>(HidController::NPad)
-        .SetSupportedStyleSet({parameters.supported_styleset});
+    GetAppletResource()
+        ->GetController<Controller_NPad>(HidController::NPad)
+        .ApplyNpadSystemCommonPolicy();
```
Comment (german77) [inline src/core/hle/service/hid/hid.cpp:1010]: This is a hack. Remove this change.

File: src/yuzu/main.cpp [hunk 14]
```diff
@@ -4346,6 +4346,30 @@ void GMainWindow::OnMiiEdit() {
     BootGame(filename, MiiEditId);
 }
 
+void GMainWindow::OnOpenControllerMenu() {
+    constexpr u64 ControllerAppletId = 0x0100000000001003ull;
```
Comment (german77) [inline src/yuzu/main.cpp:4350]: We now have an enum for this value

File: src/core/hle/service/hid/hid.cpp [hunk 15]
```diff
@@ -2802,6 +2970,70 @@ class HidSys final : public ServiceFramework<HidSys> {
...
+        rb.Push(ResultSuccess);
+        rb.PushCopyObjects(acquire_device_registered_event->GetReadableEvent());
+    }
+
+    void GetRegisteredDevices(HLERequestContext& ctx) {
+        LOG_WARNING(Service_HID, "(STUBBED) called");
+
+        struct RegisterData {
+            std::array<u8, 0x68> data;
+        };
```
Comment (german77) [inline src/core/hle/service/hid/hid.cpp:2986]: Assert on size

File: src/core/hle/service/am/am.cpp (hunk 7 above)
Comment (Macj0rdan) [reply src/core/hle/service/am/am.cpp:1830]: Odd not sure how that got here

File: src/core/hle/service/am/applets/applet_controller.h (hunk 9 above)
Comment (Macj0rdan) [reply src/core/hle/service/am/applets/applet_controller.h:37]: Yeah that was for me. Forgot to remove that.

File: src/yuzu/main.cpp [hunk 16]
```diff
@@ -4346,6 +4346,31 @@ void GMainWindow::OnMiiEdit() {
...
+        QMessageBox::warning(this, tr("Controller Applet"),
+                             tr("Controller Menu is not available. Please reinstall firmware."));
+        return;
+    }
+
+    system->GetAppletManager().SetCurrentAppletId(Service::AM::Applets::AppletId::Controller);
+
+    const auto filename = QString::fromStdString((controller_applet_nca->GetFullPath()));
+    UISettings::values.roms_path = QFileInfo(filename).path();
+    BootGame(filename);
```
Comment (german77) [inline src/yuzu/main.cpp:4371]: BootGame(filename, ControllerAppletId);

**Comment by Macj0rdan** [comment]
One issue came up. Fixing one of the hacks lead to no controller being supported. It can be merged but I will have to fix that.

File: src/core/hle/service/am/am.cpp [hunk 17]
```diff
@@ -97,6 +101,16 @@ void IWindowController::GetAppletResourceUserId(HLERequestContext& ctx) {
     rb.Push<u64>(process_id);
 }
 
+void IWindowController::GetAppletResourceUserIdOfCallerApplet(HLERequestContext& ctx) {
+    const u64 process_id = 0;
+
+    LOG_DEBUG(Service_AM, "(STUBBED) called");
```
Comment (FearlessTobi) [inline src/core/hle/service/am/am.cpp:107]: unless they are spammed, stubbed logs should be warnings

```suggestion
    LOG_WARNING(Service_AM, "(STUBBED) called");
```

File: src/core/hle/service/am/am.cpp (hunk 2 above)
Comment (FearlessTobi) [reply src/core/hle/service/am/am.cpp:1691]: This was marked as resolved but is not fixed yet.

File: src/core/hle/service/am/am.cpp [hunk 18]
```diff
@@ -1737,6 +1773,57 @@ void ILibraryAppletSelfAccessor::PushInShowAlbum() {
     queue_data.emplace_back(std::move(settings_data));
 }
 
+void ILibraryAppletSelfAccessor::PushInShowController() {
+    const auto hid = system.ServiceManager().GetService<Service::HID::Hid>("hid");
+    const auto applet_resource = hid->GetAppletResource();
+    Applets::CommonArguments common_args = {
```
Comment (FearlessTobi) [inline src/core/hle/service/am/am.cpp:1779]:
```suggestion
    const Applets::CommonArguments common_args = {
```
dito below

File: src/core/hle/service/btm/btm.cpp [hunk 19]
```diff
@@ -280,18 +280,58 @@ class IBtmSystemCore final : public ServiceFramework<IBtmSystemCore> {
...
+            {23, &IBtmSystemCore::CancelAudioDeviceConnectionRejection, "CancelAudioDeviceConnectionRejection"}
         };
         // clang-format on
 
         RegisterHandlers(functions);
     }
+
+private:
+    void IsRadioEnabled(HLERequestContext& ctx) {
+        LOG_DEBUG(Service_BTM, "(STUBBED) called");
```
Comment (FearlessTobi) [inline src/core/hle/service/btm/btm.cpp:298]: same here with warning logging

File: src/core/hle/service/btm/btm.cpp [hunk 20]
```diff
@@ -280,18 +280,58 @@ class IBtmSystemCore final : public ServiceFramework<IBtmSystemCore> {
...
+
+        IPC::ResponseBuilder rb{ctx, 3};
+        rb.Push(ResultSuccess);
+        rb.Push(1);
```
Comment (FearlessTobi) [inline src/core/hle/service/btm/btm.cpp:302]: ` rb.Push(true)`?

File: src/core/hle/service/ldn/ldn.cpp [hunk 21]
```diff
@@ -115,12 +115,20 @@ class ISystemLocalCommunicationService final
...
+            {403, &ISystemLocalCommunicationService::InitializeSystem2, "InitializeSystem2"},
         };
         // clang-format on
 
         RegisterHandlers(functions);
     }
+
+private:
+    void InitializeSystem2(HLERequestContext& ctx) {
+        LOG_DEBUG(Service_LDN, "(STUBBED) called");
```
Comment (FearlessTobi) [inline src/core/hle/service/ldn/ldn.cpp:127]: dito. I'm not going to comment this again, but there is a few more cases of this in other files.

File: src/core/hle/service/set/set_sys.cpp [hunk 22]
```diff
@@ -492,6 +491,25 @@ void SET_SYS::GetChineseTraditionalInputMethod(HLERequestContext& ctx) {
     rb.PushEnum(ChineseTraditionalInputMethod::Unknown0);
 }
 
+void SET_SYS::GetHomeMenuScheme(HLERequestContext& ctx) {
+    LOG_DEBUG(Service_SET, "(STUBBED) called");
+
+    IPC::ResponseBuilder rb{ctx, 7};
+    rb.Push(ResultSuccess);
+    rb.Push(0xFF323232);
```
Comment (FearlessTobi) [inline src/core/hle/service/set/set_sys.cpp:499]: Do we know what these numbers are?
Can we document them?

**Review by FearlessTobi — COMMENTED** [review]
Thanks for adding this!
A few comments

File: src/core/hle/service/am/am.cpp (hunk 17 above)
Comment (Macj0rdan) [reply src/core/hle/service/am/am.cpp:107]: Narr already told me this.
I will check again just to be sure but most of the lines are big spammers when the controller applet is open.

**Comment by Macj0rdan** [comment]
Working on fixing the merging issues

File: src/core/hle/service/set/set_sys.cpp (hunk 22 above)
Comment (Macj0rdan) [reply src/core/hle/service/set/set_sys.cpp:499]: This function came from Narr. @german77 do you know what these values mean?

**Comment by Macj0rdan** [comment]
I rechecked the log calls and changed some to warning and left comments for the biggest spammers

File: src/core/hle/service/set/set_sys.cpp (hunk 22 above)
Comment (german77) [reply src/core/hle/service/set/set_sys.cpp:499]: They are colors https://switchbrew.org/wiki/Settings_services#GetHomeMenuScheme you can probably make a struct

File: src/core/hle/service/set/set_sys.h [hunk 23]
```diff
@@ -269,6 +269,16 @@ class SET_SYS final : public ServiceFramework<SET_SYS> {
     };
     static_assert(sizeof(EulaVersion) == 0x30, "EulaVersion is incorrect size");
 
+    /// This is nn::settings::system::HomeMenuScheme
+    struct HomeMenuScheme {
+        u32 Main;
```
Comment (FearlessTobi) [inline src/core/hle/service/set/set_sys.h:274]:
```suggestion
        u32 main;
```
Same for the other struct members

File: src/core/hle/service/set/set_sys.cpp [hunk 24]
```diff
@@ -492,6 +491,29 @@ void SET_SYS::GetChineseTraditionalInputMethod(HLERequestContext& ctx) {
     rb.PushEnum(ChineseTraditionalInputMethod::Unknown0);
 }
 
+void SET_SYS::GetHomeMenuScheme(HLERequestContext& ctx) {
+    LOG_DEBUG(Service_SET, "(STUBBED) called");
+
+    const HomeMenuScheme default_color_struct = {
```
Comment (FearlessTobi) [inline src/core/hle/service/set/set_sys.cpp:497]:
```suggestion
    const HomeMenuScheme default_color = {
```

**Review by FearlessTobi — APPROVED** [review]
Looks very good now, only two more things:
